### PR TITLE
Wojciech Rzeźniczak 209450 - dekorowanie driverów za pomocą observera

### PR DESCRIPTION
**Pull request by WojRz**

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package edu.kis.powp.jobs2d.drivers;
+
+import edu.kis.powp.jobs2d.Job2dDriver;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashSet;
+
+public class DriverDecorator {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java]: DriverDecoratorApplicator

File: src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    public boolean isStateToDecorate() {
+        return stateToDecorate;
+    }
+
+    public void setStateToDecorate(boolean stateToDecorate) {
+        this.stateToDecorate = stateToDecorate;
+    }
+
+    public Job2dDriver decorate(Job2dDriver job2dDriver){
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java]: change to private and add public method applyDecorationIfStateToDecorate()

File: src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        try {
+            job2dDriver = (Job2dDriver)  decoratorClass.getConstructor(Job2dDriver.class).newInstance(job2dDriver);
+        } catch (Exception e){
+            return job2dDriver;
+        }
+
+        return job2dDriver;
+    }
+
+    public Job2dDriver undoDecorate(Job2dDriver job2dDriver){
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecorator.java]: fix if necessary

File: src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java [hunk 4]
```diff
@@ -71,8 +71,9 @@ private static void setupDrivers(Application application) throws FileNotFoundExc
 
 		DrawPanelController drawerController = DrawerFeature.getDrawerController();
 		//Job2dDriver driver = new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic");
-		Job2dDriver driver = new LineDriverAdapterUseControl(new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic"));
+		Job2dDriver driver = new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic");
 		DriverFeature.addDriver("Line Simulator", driver);
+		DriverFeature.addDriverDecorator("Use Control", LineDriverAdapterUseControl.class);
 		DriverFeature.getDriverManager().setCurrentDriver(driver);
 
 		driver = new LineDriverAdapterUseControl(new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special"));
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java]: remove fixed decorator

File: src/main/java/edu/kis/powp/jobs2d/features/DriverFeature.java [hunk 5]
```diff
@@ -52,4 +69,15 @@ public static void updateDriverInfo() {
...
+	public static void checkDecorator(){
+		Job2dDriver currentDriver = driverManager.getCurrentDriver();
+
+		for (DriverDecorator driverDecorator:
+				listOfDecorators) {
+			if(driverDecorator.isStateToDecorate()) currentDriver = driverDecorator.decorate(currentDriver);
+			else currentDriver = driverDecorator.undoDecorate(currentDriver);
+
+		}
+
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/features/DriverFeature.java:69]: driverManager.setCurrentDriver (test and fix loop feedback)

**Review by kowallus — CHANGES_REQUESTED** [review]
Not bad, but probably does not work.

**Comment by WojRz** [comment]
Poprawione

File: src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecoratorManager.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package edu.kis.powp.jobs2d.drivers;
+
+import edu.kis.powp.jobs2d.Job2dDriver;
+import edu.kis.powp.observer.Publisher;
+
+import java.util.ArrayList;
+
+public class DriverDecoratorManager extends Publisher {
+
+    private ArrayList<DriverDecoratorApplicator> listOfDecorators = new ArrayList<>();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/DriverDecoratorManager.java:10]: Type of attribute should be List (implementation is secondary)

**Review by kowallus — COMMENTED** [review]
OK. Final lab grade: 3

**Pull request merged**